### PR TITLE
Support for YOLOv8 based models

### PR DESCRIPTION
**Pull request by tux2000**
Using the ultralytics library directly.

Use case for me is a custom model that I trained based on the YOLOv8 model and only afterwards realized that thats not supported (yet) in doods2.

In my hands, this is working locally using the web UI. End goal would be to expose it to home assistant through your fantastic hassio add-on, which obviously is much easier if this gets to "mainline"...

### CODE REVIEW

**Comment by Sammy1Am** [comment]
Wow, just yesterday I coincidentally started to build something similar to `doods2` for YOLOv8, but then today noticed there's a PR here already!  How convenient 😄

I did notice [this commit](https://github.com/turley/doods2/commit/ba4e597e191a06ecbeba51ee3b1e43391c87379c) over in @turley 's repository though and wanted to ask:

- You've labeled this detector as `YOLOv8` versus turley's `Ultralytics`.  Is there actually anything YOLOv8-specific about this detector, or could I use it to test out e.g. YOLO-NAS or YOLOv9?  If it's not v8-specific, would it be reasonable to rename yours to `Ultralytics`?
- I don't understand the code as well as the two of you since I haven't been working with it, but as far as I can tell your changes are functionally equivalent with the exception of the inclusion of `halfPrecision` for "old AMD GPUs".  Is it reasonable to try to include that config just in case?  I'm not trying to PR-snipe turley or anything, but your changes are more up-to-date and active.

(My end goal is to add a doods2 detector type to Frigate so I can use doods2 as a backend for object detection there since it has better model support)

**Comment by Sammy1Am** [comment]
Alright, I'm having some trouble getting this to run.  I needed to add:

``` python
import odrpc
import logging
import torch
```

to yolov8.py to get it to start trying to load the detector.  But I'm trying to load a local model like this:

``` yaml
detectors:
    - name: yolov8s
      type: yolov8
      modelFile: models/yolov8s-oiv7_full_integer_quant_edgetpu.tflite
      labelFile: models/oiv7_labels.txt
```

And it's not working.  First because it seems to be trying to load the file from Torch Hub (and doesn't support local files), so I modified that and now it's having some issues with the labels.

How is your config set up that it's working?

**Comment by snowzach** [comment]
Do you think it should just be called yolo instead?

**Review by snowzach — APPROVED** [review]
I may switch it to Yolo

**Comment by Sammy1Am** [comment]
They technically have other models available as well (like FastSAM and RT-DETR): https://docs.ultralytics.com/models/

Since this detector is specifically using the `ultralytics` library (module?) to load/interact with the models, I think it should probably end up having support for anything that library supports.

If there ends up being yolo-specific optimizations at some point though, maybe `yolo` is a better name.  I guess it'll still support non-yolo models even if that's it's name.

**Comment by Sammy1Am** [comment]
Oh, also, thanks for the merge!  I'll probably have a PR tonight if I can get it to work on my machine.

**Comment by tux2000** [comment]
> I did notice [this commit](https://github.com/turley/doods2/commit/ba4e597e191a06ecbeba51ee3b1e43391c87379c) over in @turley 's repository though and wanted to ask:

I wasnt aware of @turley 's work on this.

> * You've labeled this detector as `YOLOv8` versus turley's `Ultralytics`.  Is there actually anything YOLOv8-specific about this detector, or could I use it to test out e.g. YOLO-NAS or YOLOv9?  If it's not v8-specific, would it be reasonable to rename yours to `Ultralytics`?

Sure, my intention was to separate it from the YOLOv5 support through pytorch, but YOLO or Ultralytics make sense to me too...

>  * I don't understand the code as well as the two of you since I haven't been working with it, but as far as I can tell your changes are functionally equivalent with the exception of the inclusion of `halfPrecision` for "old AMD GPUs".  Is it reasonable to try to include that config just in case?  I'm not trying to PR-snipe turley or anything, but your changes are more up-to-date and active.
> 

I went with a MVP approach to get the functionality in place that I needed for my usecase. Aditional functionality and improvements are great!

**Comment by tux2000** [comment]
> Alright, I'm having some trouble getting this to run. I needed to add:
> 
> ```python
> import odrpc
> import logging
> import torch
> ```
> 
> to yolov8.py to get it to start trying to load the detector. But I'm trying to load a local model like this:

Thats odd, could it be that you are not using the last commit? I had pushed a [version](https://github.com/snowzach/doods2/pull/112/commits/3601690ae78c563867a3f7694912f80a70fc1dfb) on my repo that did not import odrpc and had some usage of torch left, but the final version in this PR imports odrpc and logging and does not need the torch import...

> 
> ```yaml
> detectors:
>     - name: yolov8s
>       type: yolov8
>       modelFile: models/yolov8s-oiv7_full_integer_quant_edgetpu.tflite
>       labelFile: models/oiv7_labels.txt
> ```

This is the config I used:
```yaml
detectors:
    - name: yolov8
      type: yolov8
      modelFile: /share/best_20240721.pt
```
i.e. a local .pt file and no separate labels file

**Comment by tux2000** [comment]
> I may switch it to Yolo

Whatever makes most sense to you!

**Comment by Sammy1Am** [comment]
> I had pushed a version on my repo that did not import odrpc and had some usage of torch left

That must have been it, because I checked out the code after the merge and it runs just as expected!  Thanks for your work on this!

(I [made some changes](https://github.com/snowzach/doods2/compare/master...Sammy1Am:doods2:master) to add tflite support, but my Internet went out while I was writing up the PR...  Hopefully in the morning.)

**Comment by snowzach** [comment]
I rebuilt the latest docker images... I just called the detector YOLO.

**Comment by tux2000** [comment]
Great! will test. Is there any convenient way to force the hassio addon to be rebuild based on the new doods image?

**Comment by tux2000** [comment]
Ok... On the plugin page in home assistant, there is a very convenient "Rebuild" button... 
That I found after fiddling around with the commandline and pruning 30 GB of old docker images and its cache...

Now my custom YOLOv8 works flawlessly with doods2 in home assistant.

Thanks for the quick response and for making the doods2 framework available!
